### PR TITLE
Yield op tests

### PR DESCRIPTION
**Pull request by ahundt**
Add a first set of yield op tests where labels are a yield op too. These tests can be used for integration testing of new changes. 

- This PR first depends on #7060 to pass
- #7072 & deps should also cause this PR to pass.
- small part of #6928 and #7046 

Performance:
 - Note that this appears to run most efficiently in #6928, which has the `labels` parameter
    - the equivalent commit in #6928 is 9fae1b2193f006042ed6b7e0bef5d24ee4277b9e
 - It also runs in #7046, (equivalently #7072) but it may be slower.

Note: 9b4cb87 can be separated, because it tests additional cases, callbacks in particular, which not yet supported in master or #7046/#7072 with yield ops, but **callbacks are generally supported in #6928 (Yield Op support with `labels` param)**.

### CODE REVIEW

File: tests/keras/engine/test_training.py [hunk 1]
```diff
@@ -1,17 +1,22 @@
...
+from keras.layers import Dropout
+from keras.layers import Flatten
 from keras.engine.topology import Input
 from keras.engine.training import Model, _check_loss_and_target_compatibility
 from keras.models import Sequential
 from keras import backend as K
 from keras.utils import Sequence
 from keras.utils.test_utils import keras_test
 from keras.callbacks import LambdaCallback
+from keras.callbacks import TensorBoard
```
Comment (cpaxton) [inline tests/keras/engine/test_training.py]: should we include TF-specific code here?

File: tests/keras/engine/test_training.py [hunk 2]
```diff
@@ -27,6 +32,39 @@ def __getitem__(self, idx):
             np.random.random((self.batch_size, 3))]
 
 
+def call_model_methods(model, input_np, output_np, batch_size=10, epochs=1):
+
+    # train_on_batch
+    out = model.train_on_batch(input_np,
```
Comment (cpaxton) [inline tests/keras/engine/test_training.py]: don't need to set "out" here

File: tests/keras/engine/test_training.py [hunk 3]
```diff
@@ -27,6 +32,39 @@ def __getitem__(self, idx):
...
+                    output_np, epochs=1, batch_size=batch_size)
+
+    # evaluate
+    out = model.evaluate(input_np,
+                         output_np, batch_size=batch_size)
+    out = model.evaluate(input_np,
+                         output_np, batch_size=batch_size)
+
+    # predict
+    out = model.predict(input_np, batch_size=batch_size)
```
Comment (cpaxton) [inline tests/keras/engine/test_training.py]: why twice? both are not returned

File: tests/keras/engine/test_training.py [hunk 4]
```diff
@@ -451,30 +501,8 @@ def test_model_with_input_feed_tensor():
                   loss_weights=loss_weights,
                   sample_weight_mode=None)
 
-    # test train_on_batch
-    out = model.train_on_batch(input_b_np,
```
Comment (cpaxton) [inline tests/keras/engine/test_training.py]: deleted tests do not seem to be the same as new tests above -- why?

File: tests/keras/engine/test_training.py [hunk 5]
```diff
@@ -619,17 +803,11 @@ def test_model_with_external_loss():
 
     optimizer = 'rmsprop'
     loss = None
-    model.compile(optimizer, loss, metrics=['mae'])
+    model.compile(optimizer, loss, metrics=['mse'])
```
Comment (cpaxton) [inline tests/keras/engine/test_training.py]: why

File: tests/keras/engine/test_training.py [hunk 6]
```diff
@@ -530,37 +531,218 @@ def test_model_with_input_feed_tensor():
...
+
+            images = tf.parallel_stack(images, 0)
+            labels = tf.parallel_stack(labels, 0)
+            images = tf.cast(images, tf.float32)
+
+            images = tf.reshape(images, shape=batch_shape)
+
+            return images, labels
+
+    def replace(filename):
```
Comment (cpaxton) [inline tests/keras/engine/test_training.py]: why write to disk during a test?

File: tests/keras/engine/test_training.py [hunk 7]
```diff
@@ -594,6 +776,8 @@ def test_model_with_partial_loss():
     loss = {'dense_2': 'mse'}
     model.compile(optimizer, loss, metrics={'dense_1': 'mae'})
 
+    out = call_model_methods(model, input_a_np, output_a_np)
```
Comment (cpaxton) [inline tests/keras/engine/test_training.py]: What is the point to assigning to out so many times -- we don't see it anywhere else

File: tests/keras/engine/test_training.py (hunk 4 above)
Comment (ahundt) [inline tests/keras/engine/test_training.py]: I think they're extremely similar no? The new one is parameterized, so that's one difference. The prior versions aren't 100% consistent in what they test either. In other words some tests that should exist are missed with current master. However perhaps simply adding back the deletions would help merging go more quickly.

File: tests/keras/engine/test_training.py (hunk 6 above)
Comment (ahundt) [inline tests/keras/engine/test_training.py]: To test that functionality, disk tests are in many other places too like get_file. Im not aware of a way to create a tfrecord except on disk, but I haven't looked.

File: tests/keras/engine/test_training.py (hunk 7 above)
Comment (ahundt) [inline tests/keras/engine/test_training.py]: I don't know, I was just sticking to the previous behavior.

File: tests/keras/engine/test_training.py (hunk 5 above)
Comment (ahundt) [inline tests/keras/engine/test_training.py]: I didn't realize there was mean average error. It doesn't affect the test outcome or purpose but I can put it back if needed.

File: tests/keras/engine/test_training.py (hunk 3 above)
Comment (ahundt) [inline tests/keras/engine/test_training.py]: Matching previous test code. I didn't want to remove it in case they are purposely testing the = operator

File: tests/keras/engine/test_training.py (hunk 2 above)
Comment (ahundt) [inline tests/keras/engine/test_training.py]: Matching existing tests

File: tests/keras/engine/test_training.py (hunk 1 above)
Comment (ahundt) [inline tests/keras/engine/test_training.py]: The callback instance should only be created in the tf case. Import shouldn't otherwise affect things. Moving might affect pep8

File: tests/keras/engine/test_training.py (hunk 3 above)
Comment (ahundt) [inline tests/keras/engine/test_training.py]: Also I bet predict twice ensures the first call doesn't cause something to break that won't show up until the second call.

File: tests/keras/engine/test_training.py (hunk 1 above)
Comment (ahundt) [inline tests/keras/engine/test_training.py]: I forgot, This is also to highlight a bug in callbacks with a tf data source.
